common/include/scipp/common/index_composition.h

```diff
@@ -0,0 +1,130 @@
+/// @note This function uses *strides* and not a shape, meaning strides[d]
+///       is not the extent of the array in dimension d but rather the step
+///       length in the flat index to advance one element in d.
+///       It is thus *not* complementary to extract_indices.
+template <class ForwardIt1, class ForwardIt2>
+constexpr auto flat_index_from_strides(ForwardIt1 strides_it,
+                                       const ForwardIt2 strides_end,
+                                       ForwardIt2 indices_it) noexcept {
+  std::remove_const_t<std::remove_reference_t<decltype(*strides_it)>> result =
+      0;
+  for (; strides_it != strides_end; ++strides_it, ++indices_it) {
+    result += *strides_it * *indices_it;
+  }
+  return result;
+}
+
+/// Compute the bounds of a piece of memory.
+///
+/// Given a pointer `p` to some memory, a shape, and strides, this function
+/// returns a begin and an end index such that
+/// - `p + begin` is the smallest reachable address and
+/// - `p + end` is one past the largest reachable address
+///
+/// \return A pair of indices `{begin, end}`.
+template <class ForwardIt1, class ForwardIt2>
+constexpr auto memory_bounds(ForwardIt1 shape_it, const ForwardIt1 shape_end,
+                             ForwardIt2 strides_it) noexcept {
+  scipp::index begin = 0;
+  scipp::index end = 0;
+  for (; shape_it != shape_end; ++shape_it, ++strides_it) {
+    if (*strides_it < 0)
+      begin += *shape_it * *strides_it;
+    else
+      end += *shape_it * *strides_it;
+  }
+  // For scalars the above produces begin==end because all strides == 0.
+  // But we need the one-past-the-end to describe the used memory.
+  if (begin == end)
+    ++end;
+  return std::pair{begin, end};
+}
+
+/// Extract individual indices from a flat index.
+///
+/// Let
+///     I = i_0 + l_0 * (i_1 + l_1 * (i_2 + ... (i_{n-2} + l_{n-2} * i_{n-1})))
+/// be a flat index computed from indices {i_d} and shape {l_d} in
+/// 'column-major' order. Here, this means that i_0 is the fasted moving index
+/// and i_{n-1} is slowest.
+///
+/// If I == prod_{d=0}^ndim (l_d), i.e. one element past the end,
+/// the resulting indices are i_d = 0 for d < ndim-1, i_{ndim-1} = l_{ndim-1}
+/// unless l_{ndim-1} = 0, see below.
+/// This allows setting 'end-iterators' in a well defined manner.
+/// However, the result is undefined for greater values of I.
+///
+/// Values of array elements in `indices` with d > ndim-1 are unspecified.
+///
+/// Any number of l_d maybe 0 which yields i_d = 0.
+/// Except for the one-past-the-end case described above, i_{ndim-1} = 1 if
+/// l_{ndim-1} to allow this case to be distinguishable from an index
+/// to the end.
+///
+/// @param flat_index I
+/// @param ndim n
+/// @param shape {l_d}
+/// @param indices {i_d}
+/// @note This function uses a *shape*, i.e. individual dimension sizes
+///       to encode the size of the array.
+///       It is thus *not* complementary to flat_index_from_strides.
```

Comment:
You mean "It *is* thus complementary to ...`"?

Reply:
No. The point of the comment is that `extract_indices` uses a shape but `flat_index_from_strides` uses strides. Thus you need to convert between the two if you want to invert one function with the other.

Reply:
Oh, so you "complementary" in the sense of "reverse/inverse", not "set" complementary. Can you use the less ambiguous term?